cancer_ai/rewards.py

```diff
@@ -0,0 +1,56 @@
+   
+class Rewarder():
+    def __init__(self, config=None):
+        # competition_id to (hotkey_uid, days_as_leader)
+        self.competitions_leaders = {
+            1: (1, 1),
+        }
+        self.scores = {1:0}
+
+    def calculate_score_for_winner(self, competition_id: int, hotkey_uid: int) -> tuple[float, float]:
+        num_competitions = len(self.competitions_leaders)
+        base_share = 1/num_competitions
+
+        # check if current hotkey is already a leader
+        if self.competitions_leaders[competition_id][0] == hotkey_uid:
+            days_as_leader = self.competitions_leaders[competition_id][1]
+        else:
+            days_as_leader = 0
+            self.competitions_leaders[competition_id] = (hotkey_uid, days_as_leader)
+
+        if days_as_leader > 14:
+            periods = (days_as_leader - 14) // 7
+            reduction_factor = max(0.1, 1 - 0.1 * periods)
+            final_share = base_share * reduction_factor
+            reduced_share = base_share - final_share
+            return final_share, reduced_share
+        return base_share, 0
+    
+    def update_scores(self):
+        num_competitions = len(self.competitions_leaders)
+        reduced_shares_poll = {uid: 0.0 for uid in self.scores}
+
+        # If there is only one competition, the winner takes all
+        if num_competitions == 1:
+            for curr_competition_id, (current_uid, _) in self.competitions_leaders.items():
+                self.scores[current_uid] = 1.0
+            return
+        
+        for curr_competition_id, (current_uid, _) in self.competitions_leaders.items():
+            score, reduced_share = self.calculate_score_for_winner(curr_competition_id, current_uid)
+            self.scores[current_uid] += score
+
+            if reduced_share > 0:
+                # Distribute reduced share among all competitors (including the current winner if he wins another competition)
+                distributed_share = reduced_share / (num_competitions - 1)
+                for leader_competition_id, (uid, _) in self.competitions_leaders.items():
+                    if uid != current_uid or leader_competition_id != curr_competition_id:
+                        reduced_shares_poll[uid] += distributed_share
+        
+        for uid, score in self.scores.items():
+            self.scores[uid] += reduced_shares_poll[uid]
```

Comment:
are you distributing the scores among all of the competitions? Even the one reduction was taken from?

Reply:
Yes, that was the conclusion of our discussion yesterday. The way reduced_shares_poll is used is that it will feed all of the uid except for the one **instance** of the uid for a competition its iterating through.
Example:
Comp-1 = winner_uid_1
Comp-2 = winner_uid_2
Comp-3 = winner_uid_1
In that scenario when we iterate through the first competition and we get any reduced share out of it then this reduced share will be spread evenly among all the other **competitions**, **not uids**, meaning that the half of the reduced share will go to winner_uid_2, and another half back again to winner_uid_1 (because its also a leader of Comp-3), but it will not be spreaded among the winner_uid_1 twice (he will get 50%, not 66.6%) because his current instance on Comp-1 iteration that is happening is omitted.

Reply:
please write also some tests

Reply:
Do you think we should keep it as a separate class/module or it should be directly integrated into Validator? The latter would make it more complex for automating tests since there is this all overhead with learning how the testing in the template works like.

Reply:
in the same file you can, later we'll refactor